File: typescript/protocol/bidi-session.ts
import type { Transport } from './transport.js';
import type { BiDiResponse } from './types.js';
import { ProtocolError, TimeoutError } from '../utils/errors.js';

type PendingResolver = {
  resolve: (value: BiDiResponse) => void;
  reject: (err: Error) => void;
  timer: ReturnType<typeof setTimeout>;
};

type EventHandler = (params: Record<string, unknown>) => void;

/**
 * WebDriver BiDi session over the Spider WebSocket transport.
 *
 * Used for Firefox browser connections. Mirrors CDPSession's interface
 * but speaks the BiDi protocol.
 */
export class BiDiSession {
  private nextId = 1;
  private pending = new Map<number, PendingResolver>();
  private eventHandlers = new Map<string, Set<EventHandler>>();
  private transport: Transport;
  private browsingContext: string | undefined;
  private commandTimeoutMs: number;

  constructor(transport: Transport, opts?: { commandTimeoutMs?: number }) {
    this.transport = transport;
    this.commandTimeoutMs = opts?.commandTimeoutMs ?? 30_000;
  }

  get context(): string | undefined {
    return this.browsingContext;
  }

  /** Process a raw message from the transport. Returns true if handled. */
  handleMessage(data: string): boolean {
    let msg: any;
    try {
      msg = JSON.parse(data);
    } catch {
      return false;
    }

    // Response (has "id" and "type")
    if (typeof msg.id === 'number' && typeof msg.type === 'string') {
      const pending = this.pending.get(msg.id);
      if (pending) {
        this.pending.delete(msg.id);
        clearTimeout(pending.timer);
        pending.resolve(msg as BiDiResponse);
        return true;
      }
      return false;
    }

    // Event (has "type": "event" and "method")
    if (msg.type === 'event' && typeof msg.method === 'string') {
      const handlers = this.eventHandlers.get(msg.method);
      if (handlers) {
        const params = msg.params ?? {};
        for (const h of handlers) {
          try {
            h(params);
          } catch {
            // ignore
          }
        }
      }
      return true;
    }

    return false;
  }

  /** Send a BiDi command and wait for the response. */
  async send(method: string, params: Record<string, unknown>): Promise<BiDiResponse> {
    const id = this.nextId++;
    const cmd = { id, method, params };

    return new Promise<BiDiResponse>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pending.delete(id);
        reject(new TimeoutError(`BiDi command timeout: ${method} (${this.commandTimeoutMs}ms)`));
      }, this.commandTimeoutMs);

      this.pending.set(id, { resolve, reject, timer });
      try {
        this.transport.send(JSON.stringify(cmd));
      } catch (err) {
        this.pending.delete(id);
        clearTimeout(timer);
        reject(err);
      }
    });
  }

  /** Subscribe to a BiDi event. */
  on(method: string, handler: EventHandler): void {
    let set = this.eventHandlers.get(method);
    if (!set) {
      set = new Set();
      this.eventHandlers.set(method, set);
    }
    set.add(handler);
  }

  /** Unsubscribe from a BiDi event. */
  off(method: string, handler: EventHandler): void {
    this.eventHandlers.get(method)?.delete(handler);
  }

  // -------------------------------------------------------------------
  // High-level BiDi commands
  // -------------------------------------------------------------------

  /**
   * Get or create a browsing context.
   *
   * The server's Firefox relay proxies to geckodriver which already has
   * a session with a browsing context. We try multiple strategies:
   * 1. browsingContext.getTree (standard BiDi)
   * 2. session.status (to get session info, then extract context)
   * 3. browsingContext.create as last resort
   *
   * If all discovery fails, we set a placeholder context that gets replaced
   * on the first successful navigate response.
   */
  async getOrCreateContext(): Promise<string> {
    if (this.browsingContext) return this.browsingContext;

    // Strategy 1: Try browsingContext.getTree with a short timeout
    try {
      const savedTimeout = this.commandTimeoutMs;
      try {
        this.commandTimeoutMs = 5000;
        const resp = await this.send('browsingContext.getTree', {});
        const contexts = (resp.result as any)?.contexts;
        if (Array.isArray(contexts) && contexts.length > 0) {
          this.browsingContext = contexts[0].context;
          return this.browsingContext!;
        }
      } finally {
        this.commandTimeoutMs = savedTimeout;
      }
    } catch {
      // getTree not supported or timed out — try other strategies
    }

    // Strategy 2: Try creating a new context
    try {
      const savedTimeout = this.commandTimeoutMs;
      try {
        this.commandTimeoutMs = 5000;
        const createResp = await this.send('browsingContext.create', { type: 'tab' });
        const ctx = (createResp.result as any)?.context;
        if (ctx) {
          this.browsingContext = ctx;
          return this.browsingContext!;
        }
      } finally {
        this.commandTimeoutMs = savedTimeout;
      }
    } catch {
      // create not supported either
    }

    // Strategy 3: Use a placeholder — the geckodriver session already has a context,
    // and browsingContext.navigate/script.evaluate commands will use it implicitly.
    // We'll extract the real context ID from the first navigation response.
    this.browsingContext = '__default__';
    return this.browsingContext;
  }

  /** Set the browsing context (called when we discover it from a response). */
  setContext(contextId: string): void {
    this.browsingContext = contextId;
  }

  /** Navigate to a URL. */
  async navigate(url: string): Promise<void> {
    const ctx = await this.getOrCreateContext();
    const resp = await this.send('browsingContext.navigate', {
      context: ctx,
      url,
      wait: 'complete',
    });
    // Extract real context from response if we used a placeholder
    if (this.browsingContext === '__default__') {
      const realCtx = (resp.result as any)?.navigation
        ? (resp as any)?.params?.context
        : undefined;
      if (realCtx) {
        this.browsingContext = realCtx;
      }
    }
  }

  /** Capture a screenshot as base64 PNG. */
  async captureScreenshot(): Promise<string> {
    const ctx = await this.getOrCreateContext();
    const resp = await this.send('browsingContext.captureScreenshot', { context: ctx });
    const data = (resp.result as any)?.data;
    if (typeof data !== 'string') {
      throw new ProtocolError('captureScreenshot: missing result.data');
    }
    return data;
  }

  /**
   * Evaluate JavaScript and return the value.
   *
   * BiDi script.evaluate returns results in this format:
   *   { result: { type: "string", value: "..." } }
   *   { result: { type: "number", value: 42 } }
   *   { result: { type: "boolean", value: true } }
   *   { result: { type: "object", value: [...entries...] } }
   *   { result: { type: "null" } }
   *   { result: { type: "undefined" } }
   *   { result: { type: "array", value: [...] } }
   */
  async evaluate(expression: string): Promise<unknown> {
    const ctx = await this.getOrCreateContext();
    const resp = await this.send('script.evaluate', {
      expression,
      target: { context: ctx },
      awaitPromise: false,
      resultOwnership: 'none',
    });
    if (resp.type === 'error') {
      throw new ProtocolError(`BiDi script error: ${resp.message ?? resp.error}`);
    }
    // BiDi result is nested: result.result.value or result.value depending on spec version
    const resultObj = (resp.result as any)?.result ?? resp.result;
    return this.extractBiDiValue(resultObj);
  }

  /** Extract a JS value from a BiDi remote value object. */
  private extractBiDiValue(remoteValue: any): unknown {
    if (!remoteValue) return undefined;
    const type = remoteValue.type;
    switch (type) {
      case 'undefined':
        return undefined;
      case 'null':
        return null;
      case 'string':
      case 'number':
      case 'boolean':
      case 'bigint':
        return remoteValue.value;
      case 'array':
        if (Array.isArray(remoteValue.value)) {
          return remoteValue.value.map((v: any) => this.extractBiDiValue(v));
        }
        return remoteValue.value;
      case 'object':
        if (Array.isArray(remoteValue.value)) {
          // BiDi objects come as [[key, valueObj], ...] pairs
          const obj: Record<string, unknown> = {};
          for (const entry of remoteValue.value) {
            if (Array.isArray(entry) && entry.length === 2) {
              const key = typeof entry[0] === 'string' ? entry[0] : entry[0]?.value ?? String(entry[0]);
              obj[key] = this.extractBiDiValue(entry[1]);
            }
          }
          return obj;
        }
        return remoteValue.value;
      default:
        // Fallback — return value directly if present
        return remoteValue.value ?? remoteValue;
    }
  }

  /** Get full page HTML. */
  async getHTML(): Promise<string> {
    return (await this.evaluate('document.documentElement.outerHTML')) as string;
  }

  /**
   * Perform actions (BiDi input.performActions).
   * This is the BiDi way to do mouse/keyboard events.
   */
  async performActions(actions: unknown[]): Promise<void> {
    const ctx = await this.getOrCreateContext();
    await this.send('input.performActions', {
      context: ctx,
      actions,
    });
  }

  /** Click at coordinates via BiDi input actions. */
  async clickPoint(x: number, y: number): Promise<void> {
    await this.performActions([
      {
        type: 'pointer',
        id: 'mouse',
        actions: [
          { type: 'pointerMove', x: Math.round(x), y: Math.round(y) },
          { type: 'pointerDown', button: 0 },
          { type: 'pointerUp', button: 0 },
        ],
      },
    ]);
  }

  /** Insert text via BiDi input actions. */
  async insertText(text: string): Promise<void> {
    const actions = text.split('').map((ch) => [
      { type: 'keyDown', value: ch },
      { type: 'keyUp', value: ch },
    ]).flat();
    await this.performActions([{ type: 'key', id: 'keyboard', actions }]);
  }

  /** Clean up all pending commands and event handlers. Rejects in-flight commands. */
  destroy(): void {
    // Snapshot and clear first to prevent re-entrant issues from reject callbacks
    const pendingSnapshot = [...this.pending.values()];
    this.pending.clear();
    this.eventHandlers.clear();
    this.browsingContext = undefined;

    for (const pending of pendingSnapshot) {
      clearTimeout(pending.timer);
      pending.reject(new Error('session destroyed'));
    }
  }
}

File: rust/src/protocol/bidi_session.rs
//! Lock-free WebDriver BiDi session over the Spider WebSocket transport.

use crate::errors::{Result, SpiderError};
use arc_swap::ArcSwap;
use dashmap::DashMap;
use serde_json::{json, Value};
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Arc;
use tokio::sync::{mpsc, oneshot};

/// Lock-free BiDi session using DashMap for pending/events, ArcSwap for context.
pub struct BiDiSession {
    next_id: AtomicU64,
    pending: Arc<DashMap<u64, oneshot::Sender<Value>>>,
    event_handlers: Arc<DashMap<String, Vec<Arc<dyn Fn(Value) + Send + Sync>>>>,
    browsing_context: ArcSwap<Option<String>>,
    timeout_ms: u64,
    send_tx: mpsc::UnboundedSender<String>,
}

impl BiDiSession {
    pub fn new(send_tx: mpsc::UnboundedSender<String>, timeout_ms: u64) -> Self {
        Self {
            next_id: AtomicU64::new(1),
            pending: Arc::new(DashMap::new()),
            event_handlers: Arc::new(DashMap::new()),
            browsing_context: ArcSwap::from_pointee(None),
            timeout_ms,
            send_tx,
        }
    }

    pub fn context(&self) -> Option<String> {
        self.browsing_context.load().as_ref().clone()
    }

    /// Process a raw message from the transport. Returns true if handled.
    pub fn handle_message(&self, data: &str) -> bool {
        let Ok(msg) = serde_json::from_str::<Value>(data) else {
            return false;
        };

        // Response (has "id" and "type")
        if msg.get("id").and_then(|v| v.as_u64()).is_some()
            && msg.get("type").and_then(|v| v.as_str()).is_some()
        {
            let id = msg["id"].as_u64().unwrap();
            if let Some((_, tx)) = self.pending.remove(&id) {
                let _ = tx.send(msg);
                return true;
            }
            return false;
        }

        // Event (has "type": "event" and "method")
        if msg.get("type").and_then(|v| v.as_str()) == Some("event") {
            if let Some(method) = msg.get("method").and_then(|v| v.as_str()) {
                let params = msg.get("params").cloned().unwrap_or(json!({}));
                if let Some(list) = self.event_handlers.get(method) {
                    let handlers = list.clone();
                    drop(list);
                    for h in &handlers {
                        h(params.clone());
                    }
                }
                return true;
            }
        }

        false
    }

    /// Send a BiDi command and wait for the response.
    pub async fn send(&self, method: &str, params: Value) -> Result<Value> {
        let id = self.next_id.fetch_add(1, Ordering::Relaxed);
        let cmd = json!({"id": id, "method": method, "params": params});

        let (tx, rx) = oneshot::channel();
        self.pending.insert(id, tx);
        self.send_tx
            .send(cmd.to_string())
            .map_err(|_| SpiderError::connection("WebSocket is not connected"))?;

        let resp = tokio::time::timeout(
            tokio::time::Duration::from_millis(self.timeout_ms),
            rx,
        )
        .await
        .map_err(|_| {
            self.pending.remove(&id);
            SpiderError::Timeout(format!("BiDi command timeout: {method} ({}ms)", self.timeout_ms))
        })?
        .map_err(|_| SpiderError::connection("BiDi response channel closed"))?;

        if resp.get("type").and_then(|v| v.as_str()) == Some("error") {
            let msg = resp.get("message").or(resp.get("error"))
                .and_then(|v| v.as_str()).unwrap_or("unknown");
            return Err(SpiderError::Protocol(format!("BiDi error: {msg}")));
        }

        Ok(resp)
    }

    pub fn on(&self, method: &str, handler: Arc<dyn Fn(Value) + Send + Sync>) {
        self.event_handlers
            .entry(method.to_string())
            .or_default()
            .push(handler);
    }

    /// Get or create a browsing context.
    pub async fn get_or_create_context(&self) -> Result<String> {
        if let Some(ctx) = self.context() {
            return Ok(ctx);
        }

        // Strategy 1: browsingContext.getTree
        if let Ok(resp) = tokio::time::timeout(
            tokio::time::Duration::from_secs(5),
            self.send("browsingContext.getTree", json!({})),
        ).await {
            if let Ok(resp) = resp {
                if let Some(contexts) = resp.get("result")
                    .and_then(|r| r.get("contexts"))
                    .and_then(|v| v.as_array())
                {
                    if let Some(first) = contexts.first() {
                        if let Some(ctx) = first.get("context").and_then(|v| v.as_str()) {
                            self.browsing_context.store(Arc::new(Some(ctx.to_string())));
                            return Ok(ctx.to_string());
                        }
                    }
                }
            }
        }

        // Strategy 2: browsingContext.create
        if let Ok(resp) = tokio::time::timeout(
            tokio::time::Duration::from_secs(5),
            self.send("browsingContext.create", json!({"type": "tab"})),
        ).await {
            if let Ok(resp) = resp {
                if let Some(ctx) = resp.get("result")
                    .and_then(|r| r.get("context"))
                    .and_then(|v| v.as_str())
                {
                    self.browsing_context.store(Arc::new(Some(ctx.to_string())));
                    return Ok(ctx.to_string());
                }
            }
        }

        // Strategy 3: placeholder
        let placeholder = "__default__".to_string();
        self.browsing_context.store(Arc::new(Some(placeholder.clone())));
        Ok(placeholder)
    }

    pub fn set_context(&self, context_id: &str) {
        self.browsing_context.store(Arc::new(Some(context_id.to_string())));
    }

    pub async fn navigate(&self, url: &str) -> Result<()> {
        let ctx = self.get_or_create_context().await?;
        let resp = self.send("browsingContext.navigate", json!({
            "context": ctx, "url": url, "wait": "complete",
        })).await?;
        // Extract real context from response if placeholder
        if ctx == "__default__" {
            if let Some(real_ctx) = resp.get("params")
                .and_then(|p| p.get("context"))
                .and_then(|v| v.as_str())
            {
                self.set_context(real_ctx);
            }
        }
        Ok(())
    }

    pub async fn capture_screenshot(&self) -> Result<String> {
        let ctx = self.get_or_create_context().await?;
        let resp = self.send("browsingContext.captureScreenshot", json!({"context": ctx})).await?;
        resp.get("result")
            .and_then(|r| r.get("data"))
            .and_then(|v| v.as_str())
            .map(|s| s.to_string())
            .ok_or_else(|| SpiderError::Protocol("captureScreenshot: missing result.data".into()))
    }

    pub async fn evaluate(&self, expression: &str) -> Result<Value> {
        let ctx = self.get_or_create_context().await?;
        let resp = self.send("script.evaluate", json!({
            "expression": expression,
            "target": {"context": ctx},
            "awaitPromise": false,
            "resultOwnership": "none",
        })).await?;

        let result_obj = resp.get("result")
            .and_then(|r| r.get("result"))
            .or_else(|| resp.get("result"))
            .cloned()
            .unwrap_or(Value::Null);
        Ok(extract_bidi_value(&result_obj))
    }

    pub async fn get_html(&self) -> Result<String> {
        let val = self.evaluate("document.documentElement.outerHTML").await?;
        Ok(val.as_str().unwrap_or("").to_string())
    }

    pub async fn perform_actions(&self, actions: Value) -> Result<()> {
        let ctx = self.get_or_create_context().await?;
        self.send("input.performActions", json!({
            "context": ctx, "actions": actions,
        })).await?;
        Ok(())
    }

    pub async fn click_point(&self, x: f64, y: f64) -> Result<()> {
        self.perform_actions(json!([{
            "type": "pointer", "id": "mouse",
            "actions": [
                {"type": "pointerMove", "x": x.round() as i64, "y": y.round() as i64},
                {"type": "pointerDown", "button": 0},
                {"type": "pointerUp", "button": 0},
            ]
        }])).await
    }

    pub async fn insert_text(&self, text: &str) -> Result<()> {
        let actions: Vec<Value> = text.chars().flat_map(|ch| {
            let s = ch.to_string();
            vec![
                json!({"type": "keyDown", "value": s}),
                json!({"type": "keyUp", "value": s}),
            ]
        }).collect();
        self.perform_actions(json!([{"type": "key", "id": "keyboard", "actions": actions}])).await
    }

    pub fn destroy(&self) {
        self.pending.clear();
        self.event_handlers.clear();
        self.browsing_context.store(Arc::new(None));
    }
}

fn extract_bidi_value(remote: &Value) -> Value {
    match remote.get("type").and_then(|v| v.as_str()) {
        Some("undefined") | Some("null") => Value::Null,
        Some("string") | Some("number") | Some("boolean") | Some("bigint") => {
            remote.get("value").cloned().unwrap_or(Value::Null)
        }
        Some("array") => {
            if let Some(arr) = remote.get("value").and_then(|v| v.as_array()) {
                Value::Array(arr.iter().map(extract_bidi_value).collect())
            } else {
                remote.get("value").cloned().unwrap_or(Value::Null)
            }
        }
        Some("object") => {
            if let Some(pairs) = remote.get("value").and_then(|v| v.as_array()) {
                let mut map = serde_json::Map::new();
                for entry in pairs {
                    if let Some(pair) = entry.as_array() {
                        if pair.len() == 2 {
                            let key = pair[0].as_str()
                                .map(|s| s.to_string())
                                .or_else(|| pair[0].get("value").and_then(|v| v.as_str()).map(|s| s.to_string()))
                                .unwrap_or_default();
                            map.insert(key, extract_bidi_value(&pair[1]));
                        }
                    }
                }
                Value::Object(map)
            } else {
                remote.get("value").cloned().unwrap_or(Value::Null)
            }
        }
        _ => remote.get("value").cloned().unwrap_or(remote.clone()),
    }
}
